python/python/bystro/supervised_ppca/gf_dropout_pt.py

```diff
@@ -0,0 +1,314 @@
+"""
+This implements Factor analysis but with variational inference to allow us
+to supervise a subset of latent variables to be predictive of an auxiliary 
+outcome. Currently only implements logistic regression but in the future 
+will be modified to allow for more general predictive outcomes.
+
+Note that in the code W is a L x p array, where L is the latent dimensionality
+and  p is the covariate dimension, for implementation convenience. However,
+mathematically it is often pxL for notational convenience. Given that the
+most insidious errors are mathematical in nature rather than coding, (as faulty
+math is difficult to detect in unit tests), our notation matches mathematics
+rather than code, specifically when naming WWT and WTW to match the Bishop 2006
+notation rather than code.
+
+
+Objects
+-------
+PPCADropout(PPCA)
+    This is PPCA with SGD but supervises a single factor to predict y.
+
+TODO
+SPPCADropout(SPCA)
+    This is SPCA with SGD but supervises a single factor. 
+
+TODO
+FactorAnalysisDropout(FactorAnalysis)
+    This is factor analysis but supervises a single factor.
+
+Methods
+-------
+None
+"""
+import numpy as np
+
+from tqdm import trange
+import torch
+from torch import nn
+from torch.distributions.multivariate_normal import MultivariateNormal
+from sklearn.decomposition import PCA  # type: ignore
+
+from bystro.supervised_ppca._misc_np import softplus_inverse_np
+from bystro.supervised_ppca.gf_generative_pt import PPCA
+
+
+def _get_projection_matrix(W_, sigma_):
+    """
+    This is currently just implemented for PPCA due to nicer formula. Will
+    modify for broader application later.
+
+    Computes the parameters for p(S|X)
+
+    Description in future to be released paper
+
+    Parameters
+    ----------
+    W_ : pt.Tensor(n_components,p)
+        The loadings
+
+    sigma_ : pt.tensor
+        Isotropic noise
+
+    Returns
+    -------
+    Proj_X : pt.tensor(n_components,p)
+        Beta such that np.dot(Proj_X, X) = E[S|X]
+
+    Cov : pt.tensor(n_components,n_components)
+        Var(S|X)
+    """
+    n_components = int(W_.shape[0])
+    eye = torch.tensor(np.eye(n_components))
+    M = torch.matmul(W_, torch.transpose(W_, 0, 1)) + sigma_ * eye
+    Proj_X = torch.linalg.solve(M, W_)
+    Cov = torch.linalg.inv(M) * sigma_
+    return Proj_X, Cov
+
+
+class PPCADropout(PPCA):
+    """
+    This implements supervised PPCA according to the paper draft that I'm
```

Comment:
This docstring is part of the public API for the class and shouldn't contain the first-person pronoun because its referent is ambiguous.  Instead: "according to Talbot et al. (in prep, 2023) 'Prospective Title Goes Here'" so that if this docstring isn't updated in a year, the reader has some hope of finding it.